### ISSUE TITLE
Should Linux distributions developers be recommended to avoid explicitly setting `IBUS_ENABLE_SYNC_MODE=0`?

### ISSUE THREAD

**Issue by NikitkoCent** [ibus/ibus#2620]
Please fill in the following items if you don't know the root cause.

Which distribution and version?: 
`Ubuntu 22.04` (most likely many others are affected too)

Which desktop environment and version?: 
`XFCE 4.16` (there're messages that `KDE`, `Cinnamon`, `Unity` are also affected)

Which session type?: 
`X11`

Which application and version?: 
All JetBrains IDEs since the version `2023.3.2`

IBus version?: 
Including, but most likely not limited to the range `1.5.22` - `1.5.28`

Issue description:
Hello! I tried to ask these questions last week in the _ibus-dev_ google group, but haven't got any response, so trying to ask here. I apologize if it breaks the community rules. So, the text below is almost just a copy of what I've asked in the _ibus-dev_.

At JetBrains, we've faced a problem that our IDEs stop responding to any keystrokes at some moment, if there is an instance of iBus running. It happens regardless there's actual input method engine installed (e.g. ibus-pinyin or ibus-hangul) or a user only has the US layout. The issue in our tracker is there: https://youtrack.jetbrains.com/issue/JBR-6456/Sudden-keyboard-death-on-Linux-using-iBus.

Our IDEs for Linux, like any other Java Swing applications, are built using raw X11 protocol (via Xlib), so the input methods subsystem communicates with iBus using the XIM protocol and iBus's X11 frontend.
Our investigation showed that the cause of the problem is that Xlib's `XFilterEvent` call (which is `XIM_FORWARD_EVENT` on iBus's side) starts to "swallow" absolutely all keyboard events and does nothing visible in response.
I've noticed that it only happens if iBus is launched with `IBUS_ENABLE_SYNC_MODE=0` (`IBUS_ENABLE_SYNC_MODE=1` fixes the issue).

Also we've noticed that iBus has stopped to use the async mode by default (i.e. `IBUS_ENABLE_SYNC_MODE` is set to `1` by default these days). All the above mentioned observations could point us to an obvious solution of our problem - just ask everybody to set  `IBUS_ENABLE_SYNC_MODE=1`. However, developers of some Linux distributions still explicitly set it to _0_ by default, e.g. Debian/Ubuntu there: https://salsa.debian.org/input-method-team/im-config/-/commit/a629ef78f581d006499eaee633c4c98bc16e07fb. They referred to some Korean input problems as a reason to do so, such as the following ones:
* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=812826
* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=838318
* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=876794

Those problems look like a stopper for us from suggesting our users to set `IBUS_ENABLE_SYNC_MODE=1`. However, they look quite old, and may have been fixed, so I have a few question:
1. Have those input problems been fixed?
2. Is there anything currently preventing the Linux distributions developers from being recommended to _avoid_ setting `IBUS_ENABLE_SYNC_MODE=0` by default?
3. Is there a way for an external application to check if ibus-daemon has been launched in the async mode?

BTW, I have 100% stable reproducer for our problem, so feel free to ask any information or help if you need.

Steps to reproduce (actually there're completely described in our issue's [description](https://youtrack.jetbrains.com/issue/JBR-6456/Sudden-keyboard-death-on-Linux-using-iBus), so below is just a copy):
1. Run `ibus-daemon` with `IBUS_ENABLE_SYNC_MODE=0` and the xim frontend: `IBUS_ENABLE_SYNC_MODE=0 ibus-daemon --xim`
2. Enable the old ui of the IDE (_Settings_ -> _Appearance & Behavior_ -> _New UI_ -> unset the _Enable new UI_ option) and set the color scheme to IntelliJ Light (_Settings_ -> _Appearance & Behavior_ -> the _Theme_ combobox) 
3. Make sure you have Registry -> `enable.combined.diff` disabled (Ctrl+Shift+A -> type `Registry` -> choose the option `Registry...` -> find and uncheck the option `enable.combined.diff`)
4. Open a project, go to full-screen mode
5. Open the diff of a commit, go to Settings and choose “Show Diff in Separate Window”, expand the appeared window and then close it ;
6. Choose a commit in the git tool window, and start spamming the combination `Ctrl-D, Esc` in a cycle ;
7. After a few dozens repetitions, **_the IDE will stop react to any key presses_**. However, mouse clicks will still work

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
Yes, if `ibus-daemon` has been restarted with `IBUS_ENABLE_SYNC_MODE=0`

Do you see any errors when you run ibus-daemon with the verbose option?:
```bash
$ IBUS_ENABLE_SYNC_MODE=0 /usr/bin/ibus-daemon --xim --verbose &
[1] 3890
$
(ibus-ui-gtk3:3896): IBUS-WARNING **: 19:43:05.449: panel.vala:280: If you launch KDE5 on xterm, export XDG _CURRENT_DESKTOP=KDE before Launch KDE5.

(ibus-ui-gtk3:3896): IBUS-WARNING **: 19:43:05.510: ibus _bus call_ sync: org.freedesktop.DBus.Properties.Get: GDBus.Error:org.freedesktop.DBus.Error.Failed: No global engine.

```

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
yes

**Comment by NikitkoCent** [comment]
@fujiwarat, thanks for the response, I'll try today. Can you also take a look at my 3 questions above, please? They are not only about iBus itself

**Comment by fujiwarat** [comment]
> 1. Have those input problems been fixed?

Probably I think those are fixed in 1.5.29 RC2 but I'm not sure if those are my known issues exactly.

> 2. Is there anything currently preventing the Linux distributions developers from being recommended to _avoid_ setting `IBUS_ENABLE_SYNC_MODE=0` by default?

XIM does not support async key events but it just accepts the queued key events without sorting the time value.
https://github.com/ibus/ibus/issues/1713

Also GTK4 no longer supports forwarding key events for control keys likes Enter, Backspace but IBus async key event process requires to use the forwarding key events.

Now it's  a good time to make all IBus process sync.

> 3. Is there a way for an external application to check if ibus-daemon has been launched in the async mode?

GTK and QT applications can check the owned `IBUS_ENABLE_SYNC_MODE` environment variable but cannot know the default value without the variable.
X11 applications comunicate with ibus-x11 via Xorg and ibus-x11 has  the `IBUS_ENABLE_SYNC_MODE` environment variable but cannot know the default value without the variable.
The default value depends on the IBus version.

**Comment by NikitkoCent** [comment]
@fujiwarat, hello! Thanks for the replies, I appreciate them.

I didn't have time to check ibus 1.5.30 last week, but today I did. Unfortunately, the ibus built and run via:
```bash
$ sh ./configure && make && sudo make install
$ IBUS_ENABLE_SYNC_MODE=0 LD_LIBRARY_PATH=/usr/local/lib /usr/local/bin/ibus-daemon --xim --verbose
```
is still affected by the issue. Moreover, **_IBUS_ENABLE_SYNC_MODE=2 is affected too_** (_IBUS_ENABLE_SYNC_MODE=1_ is still fine though).

I must say, that the ibus run those way still partially used the binaries from the default system ibus:
```
jbr-6456    4083  0.8  0.0 386392  8960 pts/0    Sl+  mrt21   0:04 /usr/local/bin/ibus-daemon --xim --verbose
jbr-6456    4088  0.0  0.0 163552  7040 pts/0    Sl+  mrt21   0:00 /usr/libexec/ibus-memconf
jbr-6456    4089  0.3  0.1 419224 32456 pts/0    Sl+  mrt21   0:01 /usr/libexec/ibus-ui-gtk3
jbr-6456    4090  0.4  0.1 269984 26988 pts/0    Sl+  mrt21   0:01 /usr/libexec/ibus-extension-gtk3
jbr-6456    4092  0.3  0.1 188068 17792 pts/0    Sl+  mrt21   0:01 /usr/local/libexec/ibus-x11 --kill-daemon
jbr-6456    4096  0.0  0.0 237304  7424 ?        Sl   mrt21   0:00 /usr/local/libexec/ibus-portal
jbr-6456    4115  0.1  0.0 163544  7424 pts/0    Sl+  mrt21   0:00 /usr/libexec/ibus-engine-simple
jbr-6456    4393  0.0  0.0 238380  8192 pts/0    Sl+  mrt21   0:00 /usr/libexec/ibus-engine-hangul --ibus
jbr-6456    4935  0.0  0.0   9216  2560 pts/4    S+   00:05   0:00 grep --color=auto ibus
```
I don't know if these components (_ibus-memconf_, _ibus-ui-gtk3_, _ibus-extension-gtk3_, _ibus-engine-simple_, _ibus-engine-hangul_) are important for valid testing, anyway I couldn't find a way to force the local ibus-daemon to launch the local versions of them.

**Comment by NikitkoCent** [comment]
@fujiwarat,
> Probably I think those are fixed in 1.5.29 RC2 but I'm not sure if those are my known issues exactly.

Ok, got it, thanks.

> GTK and QT applications can check the owned `IBUS_ENABLE_SYNC_MODE` environment variable but cannot know the default value without the variable.
X11 applications comunicate with ibus-x11 via Xorg and ibus-x11 has the `IBUS_ENABLE_SYNC_MODE` environment variable but cannot know the default value without the variable.
The default value depends on the IBus version.

Yep, this approach won't work in case of Debian/Ubuntu because they explicitly override the variable right before launch, even not respecting a value set by the user, and the new value isn't propagated to the session environment.

> XIM does not support async key events but it just accepts the queued key events without sorting the time value.
https://github.com/ibus/ibus/issues/1713
> 
> Also GTK4 no longer supports forwarding key events for control keys likes Enter, Backspace but IBus async key event process requires to use the forwarding key events.
> 
> Now it's a good time to make all IBus process sync.

Do I understand correctly that
* https://github.com/ibus/ibus/issues/1713 is a reason to avoid using `IBUS_ENABLE_SYNC_MODE=0` for X11 apps, and
* "GTK4 no longer supports forwarding key events for control keys likes Enter, Backspace" is also a reason to avoid `IBUS_ENABLE_SYNC_MODE=0` for GTK4 apps, but
* GTK2/GTK3 apps still need `IBUS_ENABLE_SYNC_MODE=0`, because otherwise they _might_ be affected by the issues like https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=812826 (which I mentioned), if they haven't been fixed yet?

**Comment by fujiwarat** [comment]
I tried Intellij IDEA 2023.3.6 but I couldn't understand your issue.

> Make sure you have Registry -> enable.combined.diff disabled (Ctrl+Shift+A -> type Registry -> choose the option Registry... -> find and uncheck the option enable.combined.diff)

I think this option is disabled by default.

> Open the diff of a commit, go to Settings and choose “Show Diff in Separate Window”, expand the appeared window and then close it ;

I could not find the option.

> Choose a commit in the git tool window, and start spamming the combination Ctrl-D, Esc in a cycle ;

I don't see any functions with `Ctrl-D' and `ESc` keys in the left git pane in my screenshot but I see `Ctrl-D` is paste the current line in the code editor pane:

![fujiwara-ideaui](https://github.com/ibus/ibus/assets/105929/aab5f822-ad02-423b-9295-9aef94da9fa9)

**Comment by fujiwarat** [comment]
> Have those input problems been fixed?

Anyway I think those bugs were fixed some years ago.
And the current IBus default is IBUS_ENABLE_SYNC_MODE=1 for X11 applications.
https://github.com/ibus/ibus/blob/main/client/x11/main.c#L1223

**Comment by NikitkoCent** [comment]
> I tried Intellij IDEA 2023.3.6 but I couldn't understand your issue.

And I'm here ready to help 😊

> > Make sure you have Registry -> enable.combined.diff disabled (Ctrl+Shift+A -> type Registry -> choose the option Registry... -> find and uncheck the option enable.combined.diff)
> 
> I think this option is disabled by default.

Then you don't need to change anything there.

> > Open the diff of a commit, go to Settings and choose “Show Diff in Separate Window”, expand the appeared window and then close it ;
> 
> I could not find the option.

It's important for the described scenario. 
I'm attaching the video showing how you can do this:

https://github.com/ibus/ibus/assets/6972448/d196d658-3dbf-4276-aae2-605eb597ed04

Now let's move on.

> > Choose a commit in the git tool window, and start spamming the combination Ctrl-D, Esc in a cycle ;
> 
> I don't see any functions with `Ctrl-D' and `ESc`keys in the left git pane in my screenshot but I see`Ctrl-D` is paste the current line in the code editor pane:
> 
> ![fujiwara-ideaui](https://private-user-images.githubusercontent.com/105929/317058843-aab5f822-ad02-423b-9295-9aef94da9fa9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3NTQ2NDgsIm5iZiI6MTcxMTc1NDM0OCwicGF0aCI6Ii8xMDU5MjkvMzE3MDU4ODQzLWFhYjVmODIyLWFkMDItNDIzYi05Mjk1LTlhZWY5NGRhOWZhOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyOVQyMzE5MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZjIwZDFiZjJiMzIzNTU1NmZiODZkNTNjNmZmNmU1MTdjZmRjNzlkMWZmZmY4OGNkMzMxMGI5YmQ2OTk2YzFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.BKTemRRpEIA7L3iAHEoX3AssQRKkMJB2IBBv9R_9N3Y)

On this screenshot I see you opened the Commit tool window, but it's now what you need - you need the Git tool window.

I've prepared a video where the bug is getting reproduced. Before I started recording, I'd installed a utility that displays each keystroke I make. You can see this at the bottom of the video.
What is happening there:
1. I'm opening the Git tool window and selecting the first commit from the history via the mouse.
2. I'm starting spamming the sequence `Ctrl-D, Esc` in cycle until ~0:45.
3. The bug is showing up at ~0:44, because the IDE stops reacting to the `Ctrl-D` and `Esc`. After that I'm showing that the IDE has actually stopped reacting to literally any key presses I'm doing. It's exactly what is the problem about. Our users have to restart the IDE or ibus when it happens, it's the only known guaranteed way to get rid of it (until the next time).


https://github.com/ibus/ibus/assets/6972448/cf3999cc-6f12-4192-810a-a0cc7569e935

**Comment by NikitkoCent** [comment]
> > Have those input problems been fixed?
> 
> Anyway I think those bugs were fixed some years ago. And the current IBus default is IBUS_ENABLE_SYNC_MODE=1 for X11 applications. https://github.com/ibus/ibus/blob/main/client/x11/main.c#L1223

Thanks, I see. The main problem here is that Debian/Ubuntu packages developers have decided to launch iBus with `IBUS_ENABLE_SYNC_MODE` explicitly set to `0`. I'd like to ask them to stop doing that, but before I need to understand if removing `IBUS_ENABLE_SYNC_MODE=0` may cause any issues these days.

**Comment by fujiwarat** [comment]
Thank you for your explanations. I understand the reproducing steps but cannot reproduce your issue with IBUS_ENABLE_SYNC_MODE=0 for ibus-x11 with the attaching [project untitled.tar.gz](https://github.com/ibus/ibus/files/14813954/untitled.tar.gz) in the XFCE desktop session.

I already explained https://github.com/ibus/ibus/issues/2620#issuecomment-1991763881
Ubuntu has configured the wrong configuration as I sometimes suggested not to have that internal patch.

The old bugs were fixed some years ago and now ibus-hangul maintainers also agreed with it.
